### ISSUE TITLE
Unresolved variable string

### ISSUE THREAD

**Issue by MinJieLiu** [reactjs/prop-types#30]
![issues](https://minjieliu.github.io/assets/issures-image/3.png)

As `react` depends on the time, can be correctly identified in webstorm🤓

**Comment by ArianeLu** [comment]
Same. Holy shit.

**Comment by Odrin** [comment]
Yep, same here. Very annoying.

**Comment by MichaelDeBoey** [comment]
Not sure if this is a problem of the module or it's a problem of Webstorm (some config) itself?

**Comment by thebuilder** [comment]
Adding the TypeScript d.ts files to a project usually helps with code completion in WebStorm. Check out [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped). Adding the React package was the only way i could get React code completion (and React.PropTypes) to work.

A `prop-types` package has just been added, try it! https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/prop-types/

**Comment by lashae** [comment]
@thebuilder I installed the package you referenced however issue still remains. (I invalidated cache and restarted Webstorm as well.)
![](http://i67.tinypic.com/9ix0kl.jpg)

**Comment by thebuilder** [comment]
@lashae Yeah, tried it myself. Seems WebStorm can't make the connection between it and the `prop-types` module. I'm no TypeScript expert, so not sure what the correct way of setting up a package like this is.

I did manage to partially fix it by wrapping the interfaces in a typescript module. This removes the WebStorm warning, but doesn't provide the code hinting.

**prop-types.d.ts**
```ts
declare module "prop-types" {
  interface Validator<T> {
    (object: T, key: string, componentName: string, ...rest: any[]): Error | null;
  }

  interface Requireable<T> extends Validator<T> {
    isRequired: Validator<T>;
  }

  type ValidationMap<T> = { [K in keyof T]?: Validator<T> };

  interface PropTypes {
    string: Requireable<any>;
    any: Requireable<any>;
    array: Requireable<any>;
    bool: Requireable<any>;
    func: Requireable<any>;
    number: Requireable<any>;
    object: Requireable<any>;
    string: Requireable<any>;
    node: Requireable<any>;
    element: Requireable<any>;
    instanceOf(expectedClass: {}): Requireable<any>;
    oneOf(types: any[]): Requireable<any>;
    oneOfType(types: Validator<any>[]): Requireable<any>;
    arrayOf(type: Validator<any>): Requireable<any>;
    objectOf(type: Validator<any>): Requireable<any>;
    shape(type: ValidationMap<any>): Requireable<any>;
  }
}
```

I'm guessing the actual problem with code completion is due to the way the React packages are distibuted - Maybe WebStorm can't handle the dev/production env switches

**Comment by lashae** [comment]
@thebuilder What do you think the best thing to do is now? Should we open an issue on Webstorm issue tracker or try to reach someone (preferable a contributer) at least to verify that this issue is due to prop-types project?

**Comment by thebuilder** [comment]
https://youtrack.jetbrains.com/issue/WEB-26418

**Comment by gaearon** [comment]
Thanks for raising! Closing as this is not actionable for us, but glad you filed the issue with WS.

**Comment by codebymikey** [comment]
Based on the issue liked by @thebuilder, this should be fixed in Webstorm 2017.1.3.

**Comment by chenop** [comment]
Webstorm 2019.2
- Settings --> Languages & Frameworks --> JavaScript --> Libraries
- Download --> search for "prop-types" --> Download and Install --> Ok
